### DISCUSSION TITLE
Defining arrays/lists in CUE and exporting as OpenAPI

### DISCUSSION THREAD

**Discussion by nieware** [cue-lang/cue#1115] [Q&A]
For the following simple CUE code which defines a `match` object containing two team names (as an array of two strings) and the score of the match (as an array of two integers):

```
#match: {
	id: int
	teams: 2 * [string]
	score: 2 * [int]
}
```

when I run `cue export --out openapi test.cue` (with CUE v0.4.0), I get the error messages:

```
components.schemas.match.properties.score.default.0: incomplete value int
components.schemas.match.properties.score.default.1: incomplete value int
components.schemas.match.properties.teams.default.0: incomplete value string
components.schemas.match.properties.teams.default.1: incomplete value string
```

If I change `string` to `string | *null`, it works - however I don't really see why the elements of an array need to have a default value in order to be exported as OpenAPI (especially since there is no such requirement for object properties)? Am I missing something here or doing something wrong?

But even if I define default values for both array elements, the generated output contains the following:

```
"teams": {
    "type": "array",
    "items": [
        {
            "$ref": "#/components/schemas/match._.x"
        },
        {
            "$ref": "#/components/schemas/match._.x"
        }
    ],
    "default": [
        null,
        null
    ]
},
```

The references under "items" look a bit strange. "match.[any/top].x" ? I'm pretty new to CUE, so I'm not sure what's going on here?

**Comment by eadlam** [comment]
@nieware This works without specifying a list length:
```
#match: {
	id: int
	teams: [...string]
	score: [...int]
}
```
produces:
```
{
    "openapi": "3.0.0",
    "info": {
        "title": "Generated by cue.",
        "version": "no version"
    },
    "paths": {},
    "components": {
        "schemas": {
            "match": {
                "type": "object",
                "required": [
                    "id",
                    "teams",
                    "score"
                ],
                "properties": {
                    "id": {
                        "type": "integer"
                    },
                    "teams": {
                        "type": "array",
                        "items": {
                            "type": "string"
                        }
                    },
                    "score": {
                        "type": "array",
                        "items": {
                            "type": "integer"
                        }
                    }
                }
            }
        }
    }
}
```

And this seems to work as you intended with defaults:
```
import "list"

let Size = 2

#match: {
	id: int
	teams: [for i in list.Range(0, Size, 1) {string | *null}]
	score: [for i in list.Range(0, Size, 1) {int | *0}]
}
```
Produces:
```
{
    "openapi": "3.0.0",
    "info": {
        "title": "Generated by cue.",
        "version": "no version"
    },
    "paths": {},
    "components": {
        "schemas": {
            "match": {
                "type": "object",
                "required": [
                    "id",
                    "teams",
                    "score"
                ],
                "properties": {
                    "id": {
                        "type": "integer"
                    },
                    "teams": {
                        "type": "array",
                        "items": [
                            {
                                "type": "string",
                                "default": null,
                                "nullable": true
                            },
                            {
                                "type": "string",
                                "default": null,
                                "nullable": true
                            }
                        ],
                        "default": [
                            null,
                            null
                        ]
                    },
                    "score": {
                        "type": "array",
                        "items": [
                            {
                                "type": "integer",
                                "default": 0
                            },
                            {
                                "type": "integer",
                                "default": 0
                            }
                        ],
                        "default": [
                            0,
                            0
                        ]
                    }
                }
            }
        }
    }
}
```

It seems there is a bug related to the list multiplication syntax (@mpvl, @myitcv).

**Reply by nieware** [reply]
Thanks for the quick and detailed answer @eadlam ! It seems there are two issues at play here:
- when using the multiplication syntax with default, you get `"$ref": "#/components/schemas/match._.x"`
- using `for` (or enumerating the array elements explicitly) _with default_ works, but not without default, like I was originally trying to do. Is there a reason why you need to specify a default value for (elements of) constant-length lists, but not for variable-length lists and objects?

This is an example which includes all test cases:
```
import "list"

let Size = 2

#testObject: {
	// Variable-length array (works without default value)
	arrayVarNoDefault: [...string]
	// Variable-length array (also works with default value)
	arrayVarDefault: [...string | *""]
	// Error "incomplete value string"
	arrayMulNoDefault: 2 * [string]
	// "$ref": "#/components/schemas/match._.x" under "items"
	arrayMulDefault: 2 * [string | *""]
	// Error "incomplete value string"
	arrayEnumNoDefault: [string, string]
	// works correctly
	arrayEnumDefault: [string | *"", string | *""]
	// Error "incomplete value string"
	arrayForNoDefault: [for i in list.Range(0, Size, 1) {string}]
	// works correctly (same output as arrayEnumDefault)
	arrayForDefault: [for i in list.Range(0, Size, 1) {string | *""}]
}
```

**Reply by myitcv** [reply]
List operators were removed back in https://github.com/cue-lang/cue/commit/172f0060cd405f30c5873b793e44300e1a3588cb. As you will see from 867f71fec51611b96fcf5bbbf2921a6c8c2bf660, https://pkg.go.dev/cuelang.org/go/pkg/list added builtins to replace that functionality. In particular you are looking for `list.Repeat` in this case.

Is this some documentation somewhere that presents the list operators as an example that needs fixing up?

**Reply by nieware** [reply]
Thanks for the clarification! For my use case, I can also write `[string, string]`, so `2*[string]` not being available anymore is not a problem. However literally the first example for lists in the tutorials (https://cuelang.org/docs/tutorials/tour/types/lists/) defines an IPv4 address as `4 * [ uint8 ]`, so that needs to be changed...

However I still have my other question about why fixed-size lists need a default value for their elements in order to be exported successfully as OpenAPI?

**Reply by myitcv** [reply]
> However literally the first example for lists in the tutorials

Thanks, fixing that in https://review.gerrithub.io/c/cue-lang/cue/+/520994

> However I still have my other question about why fixed-size lists need a default value for their elements in order to be exported successfully as OpenAPI?

A list `[...int]` defaults to the empty list. A list `[int, int]` does not, and instead raises incomplete errors for non concrete values, `int` in this case. So I think the behaviour you are seeing is as expected: either you need to specify default values for the non-optional elements of the list, or make the list itself optional.

**Reply by nieware** [reply]
I'm currently exploring CUE mostly as a more convenient way to generate OpenAPI - I know that still needs some work, but I think it can already save you a lot of time anyway. But from the perspective of an API, a list with a fixed size of two might as well be an object: 

```
[value0, value1]  →  {element0: value0, element1: value1}
```

So it's a bit surprising that I don't need default values for the object properties, but I need default values for the array elements when exporting to OpenAPI...

**Reply by myitcv** [reply]
> But from the perspective of an API, a list with a fixed size of two might as well be an object:

This is a good analogy, and raises some interesting questions in the context of #822. Will discuss with @mpvl

**Reply by jbcpollak** [reply]
> A list [...int] defaults to the empty list. A list [int, int] does not, and instead raises incomplete errors for non concrete values, int in this case. So I think the behaviour you are seeing is as expected: either you need to specify default values for the non-optional elements of the list, or make the list itself optional.

I'm confused by this statement - I'm sure I'm misunderstanding something, but OpenAPI is a schema definition, so why are defaults necessary? Its true you'd need valid defaults if you want the list to be optional, but you can describe a two-item non-optional array in CUE, as well as in OpenAPI, shouldn't you be able to generate one from the other?
